src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp

```diff
@@ -82,44 +82,72 @@ void DependentGroupByReductionRule::_apply_to_plan_without_subqueries(
     // Also store a copy of the aggregate's output expressions to verify the output column order later on.
     const auto initial_aggregate_output_expressions = aggregate_node.output_expressions();
 
     // Gather group-by columns
     const auto fetch_group_by_columns = [&aggregate_node]() {
-      ExpressionUnorderedSet group_by_columns(aggregate_node.aggregate_expressions_begin_idx + 1);
+      auto group_by_columns = ExpressionUnorderedSet{aggregate_node.aggregate_expressions_begin_idx + 1};
       auto node_expressions_iter = aggregate_node.node_expressions.cbegin();
       std::copy(node_expressions_iter, node_expressions_iter + aggregate_node.aggregate_expressions_begin_idx,
                 std::inserter(group_by_columns, group_by_columns.end()));
       return group_by_columns;
     };
     auto group_by_columns = fetch_group_by_columns();
 
+    // Get a sorted list of ColumnIDs from an FD's set of determinants.
+    const auto get_column_ids = [&](const auto& determinants) {
+      auto column_ids = std::vector<ColumnID>{};
+      column_ids.reserve(determinants.size());
+      std::transform(determinants.cbegin(), determinants.cend(), std::back_inserter(column_ids),
+                     [&](const auto& expression) {
+                       const auto column_id = find_expression_idx(*expression, initial_aggregate_output_expressions);
+                       Assert(column_id, "Could not find column " + expression->as_column_name());
+                       return *column_id;
+                     });
+      std::sort(column_ids.begin(), column_ids.end());
+      return column_ids;
+    };
+
     // Sort the FDs by their left set's column count in hope that the shortest will later form the group-by clause.
-    std::sort(fds.begin(), fds.end(), [](const auto& fd_left, const auto& fd_right) {
-      return fd_left.determinants.size() < fd_right.determinants.size();
+    auto ordered_fds = std::vector<FunctionalDependency>{fds.cbegin(), fds.cend()};
+    std::sort(ordered_fds.begin(), ordered_fds.end(), [&](const auto& fd_left, const auto& fd_right) {
+      const auto left_determinant_size = fd_left.determinants.size();
+      const auto right_determinant_size = fd_right.determinants.size();
+      if (left_determinant_size != right_determinant_size) {
+        return left_determinant_size < right_determinant_size;
+      }
+
+      // The FDs are expected to be equally useful for the rewrite. However, we have to decide on semantics here to make
+      // the order independent of the position of the FDs in the original set (which might differ due to standard
+      // library implementation details). Thus, we compare the ColumnIDs of the determinants.
+      const auto& left_column_ids = get_column_ids(fd_left.determinants);
+      const auto& right_column_ids = get_column_ids(fd_right.determinants);
+      return std::lexicographical_compare(left_column_ids.cbegin(), left_column_ids.cend(), right_column_ids.cbegin(),
```

Comment:
Can you add a comment why you use `std::lexicographical_compare`?

Reply:
Your're right, should make no difference to just use `<`.